### DISCUSSION TITLE
how to get data by slug?

### DISCUSSION THREAD

**Discussion by masumcity** [pocketbase/pocketbase#636] [Q&A] [answered]
in api view i can get one record by id but i want to get it by slug, like were slug='new_item'

below is the original example.

const record = await client.records.getOne('brands', 'RECORD_ID', {
    expand: 'some_relation'
});

i dont know there is any thing like

const record = await client.records.getOne('brands', 'slug='new_item'', {
    expand: 'some_relation'
});

**Comment by ganigeorgiev** [comment, marked as answer]
No. I'll consider adding a helper in the future, but for now if you want to fetch a single record by a custom filter, you can use `getList` with perPage set to 1:

```js
client.records.getList("brands", 1, 1, { filter: "slug='new_item'" });
```

**Reply by masumcity** [reply]
Awesome. Thanks

**Comment by karniv00l** [comment]
I'm doing the same thing right now, but I need security rules whether the record is retrieved by slug (`findOneBy()`) or not (I've mentioned it here also: https://github.com/pocketbase/pocketbase/issues/844#issuecomment-1294742040).

So far I've tried different combinations, I think the only one that works, but not very elegant is just to check `@request.query.page = "1" && @request.query.perPage = "1"`.

I think a better way to do this would be to check if the slug is present in the filter query:

```ts
// doesn't work unfortunatelly
@request.query.filter ~ 'slug = "%"'
```

**Reply by ganigeorgiev** [reply]
@karniv00l Hm, I haven't tested it, but the last filter should work if it is the only applied filter and has the same format as the string you are comparing. Optionally, you could also try manually wrapping the raw string with `%` (this is done automatically only if your string doesn't have the`%` symbol defined; in the future there will be a more sophisticated checks to ensure escape sequence(s)). Or in other words, you could try:

```
@request.query.filter ~ '%slug = "%"%'
```

**Reply by karniv00l** [reply]
I think there might be a bug when using `~` for the `@request.query.*` object, I'm looking at the logs in debug mode and this is what I see:

given:

- collection: `tunes` with slug attribute being `tuneId` (unique string)
- request: 

```
GET /api/collections/tunes/records?page=1&perPage=1&filter=tuneId%20%3D%20%22kZGP-7_uM9%22&expand=userProfile
```

- list action custom rule: `@request.query.filter ~ "hello there"`

```
[0.00ms] SELECT {{_collections}}.* FROM `_collections` WHERE (`id`='tunes') OR (`name`='tunes') LIMIT 1
[0.00ms] SELECT COUNT(DISTINCT(`tunes`.id)) FROM `tunes` WHERE ('%hello there%' LIKE '%tuneId = "kZGP-7_uM9"%') AND (COALESCE([[tunes.tuneId]], '') = COALESCE('kZGP-7_uM9', ''))
[0.00ms] SELECT `tunes`.* FROM `tunes` WHERE ('%hello there%' LIKE '%tuneId = "kZGP-7_uM9"%') AND (COALESCE([[tunes.tuneId]], '') = COALESCE('kZGP-7_uM9', '')) LIMIT 1
```

**Reply by ganigeorgiev** [reply]
@karniv00l Yes, this seems like a bug. We switch the operands if the left one is a string/number in case `'test' ~ someField` (see https://github.com/pocketbase/pocketbase/blob/master/tools/search/filter.go#L112), but that's not exactly correct and we should add a check to preserve the order if both operands are strings/numbers OR maybe just remove the extra "magic" and preserve the expression order. I'll think a little more on this and will update it in the upcoming v0.8 release.
